### DISCUSSION TITLE
Principal lost in Asp.Net MVC 5 during async call

### DISCUSSION THREAD

**Discussion by ajohnstone-ks** [MarimerLLC/csla#1750] [Questions] [answered]
I'm updating our Asp.Net MVC 5 app on .Net framework 4.7.2 from Csla 4.11.2 to Csla 5.3.0.

We have an async action method on our controller, which basically just does an async Fetch (DataPortal.FetchAsync) and then SaveAsync.  

I had forgotten to re-add Csla.AspNet.Mvc5, and after awaiting the Fetch method the principal was gone.  It's a custom principal/identity, but this is not .Net core so I'm assuming that should still work (we will migrate to ClaimsPrincipal/Identity once we are done with the Csla 5 update).

I added Csla.AspNet.Mvc5 but now instead the principal is lost as soon as the DataPortal_Fetch method is entered (so, even earlier than it was previously).  

I'm not sure where to look, since it was working earlier and the fix for this was back ported to 4.10 I believe.  Any ideas?

**Comment by rockfordlhotka** [comment, marked as answer]
It is possible that the way the `IContextManager` is initialized has changed, or that a different provider is being used. Probably not on purpose, but a possible bug.

The thing to check is what is the type of `Csla.ApplicationContext.ContextManager` after you've done the `AddCsla` and `UseCsla` configuration calls in `Startup.cs`. Once I know what type is being used we can try and troubleshoot what's happening.

**Reply by ajohnstone-ks** [reply]
I'll have to double check tomorrow, but I'm fairly sure it was using `Csla.Web.ApplicationContextManager` from the `Csla.AspNet` package.  We're not using `AddCsla` or `UseCsla` yet; I know there's a new thing that replaced the old `ConfigurationManager` but we're still using ConfigurationManager pretty much everywhere.

In the web.config `appSettings` we have these too (I assume they still get looked at?), but I wouldn't suspect there'd be anything that would cause an issue.

```
    <add key="CslaDefaultTransactionTimeoutInSeconds" value="1200" />
    <add key="CslaDefaultTransactionIsolationLevel" value="ReadCommitted" />
    <add key="CslaAuthentication" value="Custom" />
```

**Reply by rockfordlhotka** [reply]
The `CslaAuthentication` value affects how or if the principal flows from web server to app server. That setting means that the data portal will attempt to flow the web server's principal to the app server via serialization - which works as long as your principal/identity are serializable (`ClaimsPrincipal`/`ClaimsIdentity` are _only_ serializable via `MobileFormatter`).

Really though, step 1 should be to confirm exactly what type is being used for `ContextManager`.

**Reply by ajohnstone-ks** [reply]
Good point aboiut the `CslaAuthentication` value; we're not using a remote portal in this case.  I'll remove it to avoid future confusion.

RE the `ContextManager` I verified that it is `Csla.Web.ApplicationContextManager` from `Csla.Web` right up until the `DataPortal.FetchAsync`; as soon as I hit the breakpoint on the first line of my `DataPortal_Fetch` call, its changes to `Csla.Core.ApplicatoinContextManager` from `Csla` which seems very bizarre that the manager gets changed somehow.

**Reply by ajohnstone-ks** [reply]
@rockfordlhotka Well I tried changing the fetch from `private void DataPortal_Fetch(Guid id)` to `private Task DataPortal_Fetch(Guid id)` and making it return `Task.CompletedTask` (since there's no actual async in the DP_F itself) but the behavior is unchanged.  The ContextManager gets reset somehow during the `DataPortal.FetchAsync` call.

**Reply by rockfordlhotka** [reply]
Thank you, this at least gives us a good basis to figure out what's happening. This sounds familiar to me - an issue I believe I had to figure out with aspnetcore. I should have some time to look into this later today after a bunch of meetings (so I'll be happy to look at code!!).

**Reply by ajohnstone-ks** [reply]
@rockfordlhotka Much appreciated Rocky!  As it happens I just pushed up a small sample app that repos the issue!  

https://github.com/ajohnstone-ks/MvcPrincipalLost

**Reply by ajohnstone-ks** [reply]
Also, not sure if this is meaningful or not, but attempting to `Csla.ApplicationContext.ContextManager = new Csla.Web.ApplicationContextManager()` in the `Application_Start` of `global.asax.cs` will cause a `StackOverflowException`.

**Reply by rockfordlhotka** [reply]
I think @kellyethridge ran into this (or something similar) and the solution was that we added a `CslaFlowSynchronizationContext` setting value. Try setting that to `true` in your config file and see if that changes the behavior.

**Reply by ajohnstone-ks** [reply]
@rockfordlhotka Setting that value in the web.config resolves that issue!  Thanks for your help!
